x/cron/cron.go

```diff
@@ -0,0 +1,286 @@
+	rawTime := make([]byte, 8)
+	// Zero time does not need to put any data as the bytes are already set
+	// to zero.
+	if !t.IsZero() {
+		binary.BigEndian.PutUint64(rawTime, uint64(t.UnixNano()))
+	}
+	return append([]byte("_crontask:runat:"), rawTime...)
+}
+
+// Delete implements weave.Scheduler interface.
+func (s *Scheduler) Delete(db weave.KVStore, taskID []byte) error {
+	if ok, err := db.Has(taskID); err != nil {
+		return errors.Wrap(err, "has")
+	} else if !ok {
+		return errors.Wrap(errors.ErrNotFound, "no task")
+	}
+	if err := db.Delete(taskID); err != nil {
+		return errors.Wrap(err, "cannot delete")
+	}
+	return nil
+}
+
+// NewTicker returns a cron runner instance that is using given handler to
+// process all queued messages that execution time is due. All serialization is
+// done using provided marshaler.
+//
+// Always use the same marshaler for ticker and scheduler.
+func NewTicker(h weave.Handler, enc TaskMarshaler) *Ticker {
+	return &Ticker{
+		hn:      h,
+		enc:     enc,
+		results: NewTaskResultBucket(),
+	}
+}
+
+// Ticker allows to execute messages queued for future execution. It does this
+// by implementing weave.Ticker interface.
+type Ticker struct {
+	hn      weave.Handler
+	enc     TaskMarshaler
+	results orm.ModelBucket
+}
+
+var _ weave.Ticker = (*Ticker)(nil)
+
+// Tick implementes weave.Ticker interface.
+//
+// Tick can process any number of messages suitable for execution. All changes
+// are done atomically and apply only on success.
+func (t *Ticker) Tick(ctx context.Context, db store.CacheableKVStore) ([]common.KVPair, []weave.ValidatorUpdate) {
+	tags, vdiff, err := t.tick(ctx, db)
+	if err != nil {
+		// This is a hopeless state. This error is most likely due to a
+		// database issues or some other instance specific problems.
+		// This problem is unique to this instance and this operation
+		// most likely succeeded on other nodes. This means that there
+		// is no way we could continue operating as this instance is
+		// out of sync with the rest of the network.
+		failTask(err)
+	}
+	return tags, vdiff
+}
+
+// failTask is a variable so that it can be overwritten for tests.
+var failTask = func(err error) {
+	panic(fmt.Sprintf(`
+
+Asynchronous task failed.
+
+This error is most likely due to a database issues or some other instance
+specific problems. This problem is unique to this instance and this operation
+most likely succeeded on other nodes. This means that there is no way we could
+continue operating as this instance is out of sync with the rest of the
+network.
+
+%+v
+
+	`, err))
+}
+
+// tick process any number of tasks. It always returns a response and might
+// return an error. This method is similar to the Tick except it provides an
+// error. This makes it easier for the tests to check the result.
+func (t *Ticker) tick(ctx context.Context, db store.CacheableKVStore) ([]common.KVPair, []weave.ValidatorUpdate, error) {
+	var (
+		tags  []common.KVPair
+		vdiff []weave.ValidatorUpdate
```

Comment:
req: this is never set. See L206

Reply:
Should I remove the declaration and always explicitly return `nil`?